compiler/ir/backend.jvm/src/org/jetbrains/kotlin/backend/jvm/ir/IrUtils.kt

```diff
@@ -417,3 +419,25 @@ inline fun IrElement.hasChild(crossinline block: (IrElement) -> Boolean): Boolea
     }, null)
     return result
 }
+
+// Map declarations to original declarations before lowering.
+private val IrDeclaration.original: IrDeclaration
+    get() = (this as? IrAttributeContainer)?.attributeOwnerId as? IrDeclaration ?: this
+
+// Declarations in the scope of an externally visible inline function are implicitly part of the
+// public ABI of a Kotlin module. This function returns the visibility of a containing inline function
+// (determined *before* lowering), or null if the given declaration is not in the scope of an inline function.
+val IrDeclaration.inlineScopeVisibility: DescriptorVisibility?
+    get() {
+        var owner = original
+        while (true) {
+            if (owner is IrFunction && owner.isInline) {
+                return owner.visibility
+            }
+            owner = owner.parent.safeAs<IrDeclaration>()?.original ?: return null
+        }
+    }
+
+// True for declarations which are in the scope of an externally visible inline function.
+val IrDeclaration.isInPublicInlineScope: Boolean
```

Comment:
What about `internal` case?

Reply:
I don't think we can safely ignore `internal` declarations, since they need to be visible when using `-Xfriend-paths`.

Reply:
Yes, indeed. And sorry for pause in review